jdk/test/java/lang/VirtualThread/VirtualThreadPinnedEventThrows.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @summary Test parking when pinned and emitting the JFR VirtualThreadPinnedEvent throws
+ * @run junit VirtualThreadPinnedEventThrows
+ */
+
+import java.lang.ref.Reference;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.LockSupport;
+//import jdk.internal.event.VirtualThreadPinnedEvent;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
+
+class VirtualThreadPinnedEventThrows {
+
+    /**
+     * Test parking when pinned and creating the VirtualThreadPinnedEvent fails with OOME.
+     */
+    @Test
+    void testVirtualThreadPinnedEventCreateThrows() throws Exception {
+        VirtualThreadPinnedEvent.setCreateThrows(true);
+        try {
+            testParkWhenPinned();
+        } finally {
+            VirtualThreadPinnedEvent.setCreateThrows(false);
+        }
+    }
+
+    /**
+     * Test parking when pinned and VirtualThreadPinnedEvent.begin fails with OOME.
+     */
+    @Test
+    void testVirtualThreadPinnedEventBeginThrows() throws Exception {
+        VirtualThreadPinnedEvent.setBeginThrows(true);
+        try {
+            testParkWhenPinned();
+        } finally {
+            VirtualThreadPinnedEvent.setBeginThrows(false);
+        }
+    }
+
+    /**
+     * Test parking when pinned and VirtualThreadPinnedEvent.commit fails with OOME.
+     */
+    @Test
+    void testVirtualThreadPinnedEventCommitThrows() throws Exception {
+        VirtualThreadPinnedEvent.setCommitThrows(true);
```

Comment:
# Try

I am attempting to trigger a VirtualThreadPinnedEvent at the location of Thread.sleep(10); because this ensures that the virtual thread is already parked. Then, in the testVirtualThreadPinnedEventCreateThrows test, I catch the OutOfMemoryError, but I am not quite sure how to handle the final assertTrue(completed.get());.

And I am also not sure if this approach is correct, so I would like to consult with the teacher.


# file change
```
diff --git a/origin/VirtualThreadPinnedEventThrows.java b/test03/VirtualThreadPinnedEventThrows.java
index 17149b2..bd45577 100644
--- a/origin/VirtualThreadPinnedEventThrows.java
+++ b/test03/VirtualThreadPinnedEventThrows.java
@@ -1,4 +1,4 @@
-package origin;/*
+package test03;/*
  * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
@@ -44,7 +44,9 @@ class VirtualThreadPinnedEventThrows {
         VirtualThreadPinnedEvent.setCreateThrows(true);
         try {
             testParkWhenPinned();
-        } finally {
+        } catch (OutOfMemoryError e) {
+            System.out.println("Test passed: caught OutOfMemoryError");
+        }finally {
             VirtualThreadPinnedEvent.setCreateThrows(false);
         }
     }
@@ -95,6 +97,7 @@ class VirtualThreadPinnedEventThrows {
             while ((state = thread.getState()) != Thread.State.WAITING) {
                 assertTrue(state != Thread.State.TERMINATED);
                 Thread.sleep(10);
+                new VirtualThreadPinnedEvent();
             }
```